Crash when mercs confront the oppositon located in sectors M4 and M5

#### Bug description

Version Label: Stracciatella v0.23.0-git+03452b628
Build 04.12.02

| JA2S version | Game version        | Vanilla bug? | Used to work? |
| ------------ | ------------------- | ------------ | ------------- |
| v0.23.0-git+dfbd5ffa8| JA2 linux version  | no | yes |

JA2S Sci-Fi mode.

When having Meduna partly liberated, when the squad 1 enters the sectors M4 and M5 the game crashes after the mercs are positioned when they enter the sector. Tested all nearing meduna sectors and all of them work fine except the both mentioned.

The game crashes only on sectors M4 and M5. The error is the same for M4 and M5:

The logs:

     2026-08-05T09:07:36.877072177Z [INFO] externalized/DefaultContentManager.cc: JA2 Home Dir:                  '/home/needle/.ja2'
     2026-08-05T09:07:36.877086188Z [INFO] externalized/DefaultContentManager.cc: Root game resources directory: '/opt/ja2'
     2026-08-05T09:07:36.877105222Z [INFO] externalized/DefaultContentManager.cc: Extra data directory:          '/opt/ja2-straciatella.git'
     2026-08-05T09:07:36.877118767Z [INFO] externalized/DefaultContentManager.cc: Saved games directory:         '/home/needle/.ja2/SavedGames'
     2026-08-05T09:07:36.877131153Z [INFO] externalized/DefaultContentManager.cc: Temporary directory:           '/tmp/ja2-stracciatella-Dr2SzSHh'
     2026-08-05T09:07:36.877144605Z [INFO] externalized/ModPackContentManager.cc: Enabled mods                    'from-russia-with-love, generous-rebels, imp-quiz-honest-answers, kaerar-gun-pack'
     2026-08-05T09:07:37.624848371Z [INFO] game/GameLoop.cc: Version Label: Stracciatella v0.23.0-git+dfbd5ffa8
     2026-08-05T09:07:37.624877271Z [INFO] game/GameLoop.cc: Version #:     Build 04.12.02
     2026-08-05T09:07:37.634953714Z [ERROR] game/Utils/Cinematics.cc: Failed to open 'intro/splashscreen.smk': SGPFile::openInVfs: Vfs_open "intro/splashscreen.smk": entity not found
     2026-08-05T09:08:24.152423101Z [ERROR] sgp/Debug.h: Assertion failed in /opt/ja2-straciatella.git/src/game/Tactical/Vehicles.cc, line 104:

Not sure why the game DOES NOT crash for sector M3. Sector M3 has the same soldier/tank mixture like M4 and M5.

[Crash.zip](https://github.com/user-attachments/files/30738991/Crash.zip)

#### Steps to reproduce

1. Load the Crash.zip savegame
2. In the Map Screen use squad 1 to enter sector M5 or M4
3. Fast forward time
4. When the sector is entered, place mercs entry points
5. Crash

--- a/src/game/Strategic/Strategic_Movement.cc
+++ b/src/game/Strategic/Strategic_Movement.cc
@@ -60,7 +60,7 @@ static GROUP* gpPendingSimultaneousGroup = NULL;
 
 extern BOOLEAN gubNumAwareBattles;
 extern INT8 SquadMovementGroups[ ];
-extern INT8 gubVehicleMovementGroups[ ];
+extern UINT8 gubVehicleMovementGroups[];
 
 BOOLEAN gfDelayAutoResolveStart = FALSE;
 

--- a/src/game/Tactical/Vehicles.cc
+++ b/src/game/Tactical/Vehicles.cc
@@ -41,7 +41,7 @@
 #include <vector>
 
 
-INT8 gubVehicleMovementGroups[ MAX_VEHICLES ];
+UINT8 gubVehicleMovementGroups[MAX_VEHICLES];
 
 // the list of vehicle slots
 std::vector<VEHICLETYPE> pVehicleList;
@@ -74,10 +74,16 @@ void SetVehicleValuesIntoSoldierType(SOLDIERTYPE* const vs)
 
 INT32 AddVehicleToList(const SGPSector& sMap, INT16, const UINT8 ubType)
 {
+	// Garbage collection: due to a bug the vehicle list might contain
+	// tanks destroyed in another section and reach the MAX_VEHICLES limit.
+	for (auto & v : pVehicleList)
+	{
+		if (v.fDestroyed && v.sSector != sMap) v.fValid = false;
+	}
+
 	INT32 vid;
 	for (vid = 0;; ++vid)
 	{
-		Assert(pVehicleList.size() <= INT32_MAX);
 		if (vid == static_cast<INT32>(pVehicleList.size()))
 		{
 			pVehicleList.push_back(VEHICLETYPE{});
@@ -85,6 +91,7 @@ INT32 AddVehicleToList(const SGPSector& sMap, INT16, const UINT8 ubType)
 		}
 		if (!pVehicleList[vid].fValid) break;
 	}
+	Assert(pVehicleList.size() <= MAX_VEHICLES);
 	VEHICLETYPE* const v = &pVehicleList[vid];
 
 	// found a slot
@@ -811,7 +818,7 @@ void LoadVehicleMovementInfoFromSavedGameFile(HWFILE const hFile)
 	INT32 cnt;
 
 	//Load in the Squad movement id's
-	hFile->read(gubVehicleMovementGroups, sizeof(INT8) * 5);
+	hFile->read(gubVehicleMovementGroups, sizeof(UINT8) * 5);
 
 	for( cnt = 5; cnt <  MAX_VEHICLES; cnt++ )
 	{
@@ -826,14 +833,14 @@ void LoadVehicleMovementInfoFromSavedGameFile(HWFILE const hFile)
 void NewSaveVehicleMovementInfoToSavedGameFile(HWFILE const hFile)
 {
 	//Save all the vehicle movement id's
-	hFile->write(gubVehicleMovementGroups, sizeof(INT8) * MAX_VEHICLES);
+	hFile->write(gubVehicleMovementGroups);
 }
 
 
 void NewLoadVehicleMovementInfoFromSavedGameFile(HWFILE const hFile)
 {
 	//Load in the Squad movement id's
-	hFile->read(gubVehicleMovementGroups, sizeof(INT8) * MAX_VEHICLES);
+	hFile->read(gubVehicleMovementGroups);
 }
 
 

